Blog post on Broker Scale down check introduced in Strimzi 0.38 (#399)
strimzi/strimzi.github.io: _posts/2023-11-30-prevent-broker-scale-down-if-containing-paritition-replicas.md

---
layout: post
title: "Prevent broker scale down if containing partition replicas"
date: 2023-11-30
author: Shubham Rawat
---

> Apache Kafka is a platform which is designed to be scalable.

Reviewer: ```suggestion
Apache Kafka is a platform designed for scalability.
```

You can always scale in or scale out your Kafka Clusters based on your use case.

> When dealing with scaling down of cluster, you have to make sure that data across the brokers are moved/copied across the cluster.

Reviewer: ```suggestion
When scaling down a cluster, it's crucial to ensure that data across brokers is moved or copied throughout the cluster.
```

## Broker scale down check

This check makes sure that when you are scaling down your cluster, there are no partition replicas present on the broker that is going to be removed.
If partition replicas are found on the broker then the cluster operations are blocked and reconciliation fails until you revert the Kafka resource.
The check is enabled by default with Strimzi 0.38.

However, there may be scenarios where you want to bypass this blocking mechanism.
Disabling the check might be necessary on busy clusters, for example, because new topics keep generating replicas for the broker. .
This situation can indefinitely block cluster operations, even when brokers are nearly empty.
Overriding the blocking mechanism in this way has an impact:
the presence of topics on the broker being scaled down will likely cause a reconciliation failure for the Kafka cluster.

You can bypass the blocking mechanism by annotating the `Kafka` resource for the Kafka cluster by setting the `strimzi.io/skip-broker-scaledown-check` annotation to `true`:
```shell
kubectl annotate Kafka my-kafka-cluster strimzi.io/skip-broker-scaledown-check="true"
```

## Setting up the environment

> Let's spin up an cluster where we can work through some reassignment examples.

Reviewer: ```suggestion
Let's set up a cluster to work through some reassignment examples.
```

To get the Kafka cluster up and running, we will first have to install the Strimzi Cluster Operator and then deploy the `Kafka` resource.
You can refer to the [Stimzi Quickstart Guide](https://strimzi.io/docs/operators/latest/quickstart.html) for installing Strimzi.

You can install the Cluster Operator with any installation method you prefer.

Then we will deploy the Kafka cluster with Cruise Control enabled.

Example Kafka resource with Cruise Control enabled:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.6.0
    replicas: 4
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
      inter.broker.protocol.version: "3.6"
    storage:
      type: ephemeral
  zookeeper:

>     replicas: 4

Reviewer 2: Why we need 4 ZK replicas?

    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
  cruiseControl: {}
```

Once the cluster is up, we can create some topics such that we have some partition replicas asssigned to the brokers
Here is an exmaple topic configuration:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 3
  replicas: 3
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824

```

You can now check how topic partition replicas are allocated across the brokers:
```shell

> Topic: my-topic	TopicId: VcdsMY9gR1STjURMGEHrlA	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1

Author: I was doing it like this `kubectl exec -n myproject -ti my-cluster-kafka-0 /bin/bash -- bin/kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic my-topic` but at the same time I guess its better to use a seperate pod, so something like this ``kubectl exec -n myproject -ti my-pod /bin/bash -- bin/kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic my-topic`for these purposes. That would mean I have tell separately about getting a new pod up for this purpose?

> Topic: my-topic	TopicId: VcdsMY9gR1STjURMGEHrlA	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1

Reviewer 2: You can't use localhost from another pod. Try something like this:

```sh
kubectl run client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:latest-kafka-3.6.0" -- \
  sh -c "/opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic; exit 0";
```

	Topic: my-topic	Partition: 0	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
	Topic: my-topic	Partition: 1	Leader: 0	Replicas: 0,2,3	Isr: 0,2,3
	Topic: my-topic	Partition: 2	Leader: 2	Replicas: 2,3,1	Isr: 2,3,1
```

> ### Scaling down the replicas (without moving replicas from broker to be removed)

Reviewer 2: ```suggestion
### Scaling down without moving out all replicas from the broker to be removed
```

2023-11-30 11:08:12 WARN  AbstractOperator:557 - Reconciliation #150(watch) Kafka(myproject/my-cluster): Failed to reconcile
io.strimzi.operator.common.model.InvalidResourceException: Cannot scale down brokers [3] because brokers [3] are not empty
	at io.strimzi.operator.cluster.operator.assembly.KafkaReconciler.lambda$brokerScaleDownCheck$26(KafkaReconciler.java:300) ~[io.strimzi.cluster-operator-0.38.0.jar:0.38.0]
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:38) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
	at io.vertx.core.Promise.complete(Promise.java:66) ~[io.vertx.vertx-core-4.4.6.jar:4.4.6]
```
So these logs are basically telling you that the broker are not empty and therefore the reconciliation is failing and cluster operations are blocked. To get rid of this error you can revert the replica change in the Kafka resource.

> ### Scaling down the replicas (after emptying partition replicas from broker to be removed)

Reviewer 2: ```suggestion
### Scaling down after moving out all replicas from the broker to be removed
```

Let's try to scale down the broker now after emptying partition replicas from the broker to be removed.

> We can make use of the `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job. 

Reviewer 2: How? Please, don't assume any prior knowledge about how Strimzi works. You can explain this step in the blog post (which I think is better), or link to the relevant doc from a specific release.

> Doing this will make Cruise Control do all the job of rebalancing and moving the partition replicas from the broker that is going to be removed.

Reviewer: ```suggestion
Doing this will allow Cruise Control to handle the task of rebalancing and moving the partition replicas from the broker that is going to be removed.
```

Once the rebalacing is done, you can validate/check if the topics are move from broker or not.
```shell
	Topic: my-topic	Partition: 0	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
	Topic: my-topic	Partition: 1	Leader: 0	Replicas: 0,2,1	Isr: 0,2,1
	Topic: my-topic	Partition: 2	Leader: 2	Replicas: 2,0,1	Isr: 2,1,0
```

Now you can scale down the broker and, it will happen flawlessly since the broker is empty.

## What's next

Hope this blog post gave you a gist on how broker scale down check works.

> With the next releases of Strimzi, we are working to make this process more flawless by reverting the kafka replicas back to what they were if the broker scale dowm check fails so that the reconciliation doesn't fail and there is no blocking of cluster operations

Reviewer 2: ```suggestion
With the next releases of Strimzi, we will improve this process by automatically reverting `.spec.kafka.replicas` back to the original value when the broker scale down check fails. This means that the reconciliation doesn't fail and cluster operations are not blocked.
```

---
layout: post
title: "Preventing Kafka broker scale down if partition replicas are present"
date: 2023-12-11
author: Shubham Rawat
---

Apache Kafka is a platform designed for scalability.
You can always scale in or scale out your Kafka Clusters based on your use case.
When scaling down a cluster, it's crucial to ensure that data across brokers is moved or copied throughout the cluster.

> You may experience data loss if you forget to move out partition replicas from the broker being removed.

Reviewer 3: Would it be also worth mentioning that Strimzi might not be able to roll such cluster? I think that is more common than a data loss (which can also happen in some cases, so no need to remove it).

But don't worry, with Strimzi 0.38, we have introduced the broker scale down check which is going to take care of this problem.

> ## Broker scale down check

Reviewer 3: I wonder if it would be better to keep this for the end - before the "What's next" chapter. I find it a bit weird that you have a blog post introducing a new feature and you start with _how to disable it_. I would first explain the feature and at the end say _if needed, you can disable it_.

This check makes sure that, when you are scaling down your Kafka cluster, there are no partition replicas present on the broker that is going to be removed.
If partition replicas are found on the broker then the cluster operations are blocked and reconciliation fails until you revert the Kafka resource.
This is enabled by default with Strimzi 0.38.

However, there may be scenarios where you want to bypass this blocking mechanism.
For example, disabling the check might be necessary when new topics are being constantly created.
This situation can indefinitely block cluster operations, even when brokers are nearly empty.

You can bypass the blocking mechanism by annotating the `Kafka` resource for the Kafka cluster by setting the `strimzi.io/skip-broker-scaledown-check` annotation to `true`:
```shell
kubectl annotate Kafka my-cluster strimzi.io/skip-broker-scaledown-check="true"
```

## Setting up the environment

Let's set up a cluster to work through an example demonstrating this feature.

To get the Kafka cluster up and running, we will first have to install the Strimzi Cluster Operator and then deploy the `Kafka` resource.

> You can refer to the [Stimzi Quickstart Guide](https://strimzi.io/docs/operators/latest/quickstart.html) to install the operator.

Reviewer 3: The link does not work. Not sure exactly where you wanted to link.

You can install the Cluster Operator with any installation method you prefer.

> Then we will deploy the Kafka cluster with Cruise Control enabled.

Reviewer 3: Would it make sense to join these into one paragraph? Not something you have to do. But I think the actual blog post looks better that way then with several parapgraphs consisiting from one sentence only.

Example Kafka resource with Cruise Control enabled:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:

>     version: 3.6.0

Reviewer 3: Maybe for the blog post it is better to leave it without version? It will better stand the test of time.

Author: Hi Jakub, you mean removing the version field or do you mean just removing the version and something like `version: <kafka-version>`. ?

Reviewer 3: Removing it. That way it will work even when 3.6.0 is not supported anymore.

    replicas: 4
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2

>       inter.broker.protocol.version: "3.6"

Reviewer 3: Same as above, maybe you can remove this?

    storage:
      type: ephemeral
  zookeeper:
    replicas: 3
    storage:
      type: ephemeral
  entityOperator:
    topicOperator: {}
    userOperator: {}
  cruiseControl: {}
```

Once the cluster is up, we can create some topics such that we have some partition replicas assigned to the brokers.
This is a topic configuration example:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaTopic
metadata:
  name: my-topic
  labels:
    strimzi.io/cluster: my-cluster
spec:
  partitions: 3
  replicas: 3
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

You can now check how topic partition replicas are allocated across the brokers using the command:
```shell

> $ kubectl run -n myproject client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:latest-kafka-3.6.0" -- \
> sh -c "/opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic; exit 0";

Reviewer 3: Wouldn't this work as well and be easier to understand for users? You should also use a release image I guess.

```suggestion
$ kubectl run -n myproject client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:0.38.0-kafka-3.6.0" -- /opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic
```

	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
```

### Scaling down without moving out all replicas from the broker to be removed

Let's try to scale down the number of brokers by changing the `.spec.kafka.replicas` configuration from 4 to 3, while all brokers have partition replicas assigned.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.6.0
    replicas: 3 # Changes the replicas to 3
    listeners:
# ....    
```

Now you can apply the updated Kafka custom resource and check if the brokers are scaled down or not.

You check the status of the Kafka custom resource using this command:
```sh
kubectl get kafka my-cluster -n myproject -o yaml
```

Since we didn't move the replicas from the broker to be removed, the scale down fails and, as indicated by the errors given in the status of the Kafka custom resource:
```yaml
status:
  clusterId: S4kUmhqvTQegHCGXPrHe_A
  conditions:
  - lastTransitionTime: "2023-12-04T07:03:49.161878138Z"
    message: Cannot scale down brokers [3] because brokers [3] are not empty
    reason: InvalidResourceException
    status: "True"
    type: NotReady
```

> So these logs are basically telling you that broker 3 is not empty, which makes the whole reconciliation fail. You can get rid of this error by reverting `.spec.kafka.replicas` in the Kafka resource.

Reviewer 3: Would it be also worth showing a log snippet from the CO log? As people might look there as well.

### Scaling down after moving out all replicas from the broker to be removed

Now, let's first move out all replicas from the broker to be removed, before attempting a new Kafka cluster scale down.

We can make use of Cruise Control's `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job.
Here is an example `KafkaRebalance` resource:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:

>     strimzi.io/cluster: my-cluster

Reviewer 3: Should we add the auto approval annotation to make the demo included in the blog post easily reproducible?

---
layout: post
title: "Prevent broker scale down if containing partition replicas"
date: 2023-11-30
author: Shubham Rawat
---

Apache Kafka is a platform designed for scalability.
You can always scale in or scale out your Kafka Clusters based on your use case.
When scaling down a cluster, it's crucial to ensure that data across brokers is moved or copied throughout the cluster.
There can be a possibility that while scaling down, you forgot to remove the partition replicas from the broker that is going to be removed and due to that you will have to suffer data loss.
But don't worry, with Strimzi 0.38, we have introduced the broker scale down check which is going to take care of this problem.

## Broker scale down check

This check makes sure that when you are scaling down your cluster, there are no partition replicas present on the broker that is going to be removed.
If partition replicas are found on the broker then the cluster operations are blocked and reconciliation fails until you revert the Kafka resource.
The check is enabled by default with Strimzi 0.38.

However, there may be scenarios where you want to bypass this blocking mechanism.

> Disabling the check might be necessary on busy clusters, for example, because new topics keep are being created all the time on such clusters.

Author: Why do we add ``` here @PaulRMellor ?

> Disabling the check might be necessary on busy clusters, for example, because new topics keep are being created all the time on such clusters.

Reviewer 4: Ah, sorry @ShubhamRwt. It left the suggestion ticks in for some reason.

Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,3,1	Isr: 2,3,1
	Topic: my-topic	Partition: 1	Leader: 3	Replicas: 3,1,0	Isr: 3,1,0
	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
```
using the command:
```shell
kubectl run -n myproject client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:latest-kafka-3.6.0" -- \
sh -c "/opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic; exit 0";
```

### Scaling down without moving out all replicas from the broker to be removed

Let's try to scale down `.spec.kafka.replicas` (no. of brokers) from 4 to 3, while all brokers have partition replicas assigned.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.6.0
    replicas: 3            # Changes the replicas to 3
    listeners:
# ....    
```

Now you can apply the updated Kafka custom resource and check if the broker are scaled down or not.

> You check the status of the Kafka CR using this command:

Reviewer 4: ```suggestion
You check the status of the Kafka custom resource using this command:
```

```sh
kubectl get kafka my-cluster -n myproject -o yaml
```

> Since we didn't move the replicas from the broker to be removed, the scale down will fail and, you will be able to see these errors in the status of the Kafka custom resource

Reviewer 4: ```suggestion
Since we didn't move the replicas from the broker to be removed, the scale down fails and, as indicated by the errors given in the status of the Kafka custom resource:
```

```yaml
status:
  clusterId: S4kUmhqvTQegHCGXPrHe_A
  conditions:
  - lastTransitionTime: "2023-12-04T07:03:49.161878138Z"
    message: Cannot scale down brokers [3] because brokers [3] are not empty
    reason: InvalidResourceException
    status: "True"
    type: NotReady
```
So these logs are basically telling you that broker 3 is not empty, which makes the whole reconciliation fail. You can get rid of this error by reverting `.spec.kafka.replicas` in the Kafka resource.

### Scaling down after moving out all replicas from the broker to be removed

Let's try to scale down the broker now after emptying partition replicas from the broker to be removed.

> We can make use of the `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job.

Reviewer 4: ```suggestion
We can make use of Cruise Control's `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job.
```

> Here is an example `KafkaRebalance` resource.

Reviewer 4: ```suggestion
Here is an example `KafkaRebalance` resource:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
# no goals specified, using the default goals from the Cruise Control configuration
spec:
  mode: remove-brokers
  brokers: [3]
```
You can create this `KafkaRebalance` custom resource and this will allow Cruise Control to handle the task of rebalancing and moving the partition replicas from the broker that is going to be removed.
For more in-depth information you can refer to our [Rebalancing cluster using Cruise Control](https://strimzi.io/docs/operators/latest/deploying#proc-generating-optimization-proposals-str) documentation.

> Once the rebalacing is done, you can validate/check if the topics are move from broker or not.

Reviewer 4: ```suggestion
Once the rebalacing is done, you can check if the topics are moved from the broker or not:
```

```shell
Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,0,1	Isr: 2,1,0
	Topic: my-topic	Partition: 1	Leader: 2	Replicas: 2,1,0	Isr: 1,0,2
	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
```

> Now you can scale down the broker and, it will happen flawlessly since the broker is empty.

Reviewer 4: ```suggestion
Now you can scale down the broker, and it will happen seamlessly since the broker is empty.
```

## What's next

> Hope this blog post gave you a gist on how broker scale down check works.
> With the next releases of Strimzi, we will improve this process by automatically reverting `.spec.kafka.replicas` back to the original value when the broker scale down check fails. This means that the reconciliation doesn't fail and cluster operations are not blocked.

Reviewer 4: ```suggestion
We hope this blog post has provided you with a clear understanding of how the broker scale-down check operates. 
In upcoming Strimzi releases, we aim to enhance this process by automatically reverting `.spec.kafka.replicas` to its original value if the broker scale-down check fails. 
This improvement ensures that reconciliation doesn't fail, and cluster operations remain unblocked.
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 3.6.0
    replicas: 3 # Changes the replicas to 3
    listeners:
# ....    
```

Now you can apply the updated Kafka custom resource and check if the brokers are scaled down or not.

You check the status of the Kafka custom resource using this command:
```sh
kubectl get kafka my-cluster -n myproject -o yaml
```

Since we didn't move the replicas from the broker to be removed, the scale down fails and, as indicated by the errors given in the status of the Kafka custom resource:
```yaml
status:
  clusterId: S4kUmhqvTQegHCGXPrHe_A
  conditions:
  - lastTransitionTime: "2023-12-04T07:03:49.161878138Z"
    message: Cannot scale down brokers [3] because brokers [3] are not empty
    reason: InvalidResourceException
    status: "True"
    type: NotReady
```
So these logs are basically telling you that broker 3 is not empty, which makes the whole reconciliation fail. You can get rid of this error by reverting `.spec.kafka.replicas` in the Kafka resource.

### Scaling down after moving out all replicas from the broker to be removed

> Let's try to scale down the broker now after emptying partition replicas from the broker to be removed.

Reviewer 2: ```suggestion
Now, let's first move out all replicas from the broker to be removed, before attempting a new Kafka cluster scale down.
```

spec:
  mode: remove-brokers
  brokers: [3]
```
You can create this `KafkaRebalance` custom resource and this will allow Cruise Control to handle the task of rebalancing and moving the partition replicas from the broker that is going to be removed.
For more in-depth information you can refer to our [Rebalancing cluster using Cruise Control](https://strimzi.io/docs/operators/latest/deploying#proc-generating-optimization-proposals-str) documentation.

Once the rebalacing is done, you can check if the topics are moved from the broker or not:
```shell
Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,0,1	Isr: 2,1,0
	Topic: my-topic	Partition: 1	Leader: 2	Replicas: 2,1,0	Isr: 1,0,2
	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
```

Now you can scale down the broker, and it will happen seamlessly since the broker is empty.

## What's next

We hope this blog post has provided you with a clear understanding of how the broker scale-down check operates.

> In upcoming Strimzi releases, we aim to enhance this process by automatically reverting `.spec.kafka.replicas` to its original value if the broker scale-down check fails.

Reviewer 2: Is there general agreement on that? I don't remember any place where we update the spec, but maybe I'm wrong.

Reviewer 3: This is a good point from Federico ...
* This will likely not be in the next Strimzi release
* We do not really plan to revert the `.spec.kafka.replicas` => even if we did it is not used with KRaft. I would word it as something like _continue the reconciliation without doing the scaledown_ or something in that sense which is what we plan.

---
layout: post
title: "Preventing Kafka broker scale down if partition replicas are present"
date: 2023-12-11
author: Shubham Rawat
---

Apache Kafka is a platform designed for scalability.
You can always scale in or scale out your Kafka Clusters based on your use case.
When scaling down a cluster, it's crucial to ensure that data across brokers is moved or copied throughout the cluster.
If you forget to move out the partition replicas from the broker being removed, then Strimzi might not be able to roll the cluster properly due to availability issues i.e. replicas assigned to that broker will be always offline. 

> Another issue would be that you would suffer data loss.

Reviewer 3: ```suggestion
Another issue would be that you might suffer data loss.
```

Author: Thanks for another pass Jakub. Applied the change

But don't worry, with Strimzi 0.38, we have introduced the broker scale down check which is going to take care of this problem.

## Broker scale down check

This check makes sure that, when you are scaling down your Kafka cluster, there are no partition replicas present on the broker that is going to be removed.
If partition replicas are found on the broker then the cluster operations are blocked and reconciliation fails until you revert the Kafka resource.
This is enabled by default with Strimzi 0.38.

## Setting up the environment

Let's set up a cluster to work through an example demonstrating this feature.
To get the Kafka cluster up and running, we will first have to install the Strimzi Cluster Operator and then deploy the `Kafka` resource.

> You can refer to the [Stimzi Quickstart Guide](https://strimzi.io/quickstarts/) to install the operator.
> You can install the Cluster Operator with any installation method you prefer.

Reviewer 3: Would it make sense to reverse it?

```suggestion
You can install the Cluster Operator with any installation method you prefer.
For example by following the [Stimzi Quickstart Guide](https://strimzi.io/quickstarts/).
```

Author: Done.

spec:
  partitions: 3
  replicas: 3
  config:
    retention.ms: 7200000
    segment.bytes: 1073741824
```

You can now check how topic partition replicas are allocated across the brokers using the command:
```shell
$ kubectl run -n myproject client -itq --rm --restart="Never" --image="quay.io/strimzi/kafka:0.38.0-kafka-3.6.0" -- /opt/kafka/bin/kafka-topics.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --describe --topic my-topic

Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,3,1	Isr: 2,3,1
	Topic: my-topic	Partition: 1	Leader: 3	Replicas: 3,1,0	Isr: 3,1,0
	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
```

### Scaling down without moving out all replicas from the broker to be removed

Let's try to scale down the number of brokers by changing the `.spec.kafka.replicas` configuration from 4 to 3, while all brokers have partition replicas assigned.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:

>     version: 3.6.0

Reviewer 3: ```suggestion
```

Author: Done
